python/ray/experimental/channel/torch_tensor_nccl_channel.py

```diff
@@ -63,24 +67,62 @@ def __init__(
                 driver.
             reader_and_node_list: A list of tuples, where each tuple contains a reader
                 actor handle and the node ID where the actor is located.
-            gpu_data_channel: A GPU-GPU channel for sending tensor data. Its
-                writer and readers should match the given writer and readers.
-            cpu_data_channel: A shared-memory channel for sending
+            typ: Type information about the values passed through the channel.
+            driver_actor_id: The actor ID of the DAGDriverProxyActor.
+            tensor_metadata_channel: A shared-memory channel for sending tensor
+                metadata.
+            _cpu_data_channel: A shared-memory channel for sending
                 non-tensor data. Its writer and readers should match the given
                 writer and readers. If None is provided, then we assume that
                 there is no CPU-specific data, i.e. the task directly returned
                 a CUDA torch.Tensor.
+            _gpu_data_channel: A channel for sending torch.Tensors via NCCL.
+            _local_channel: A channel for sending data between the writer and
+                local readers.
         """
         self._writer = writer
         self._reader_and_node_list = reader_and_node_list
         self._typ = typ
 
-        self._gpu_data_channel: _TorchTensorNcclChannel = gpu_data_channel
-        if self._typ.direct_return:
-            self._cpu_data_channel = None
-        else:
-            assert cpu_data_channel is not None
-            self._cpu_data_channel: Optional["Channel"] = cpu_data_channel
+        (
+            remote_reader_and_node_list,
+            local_reader_and_node_list,
+        ) = utils.split_readers_by_locality(self._writer, self._reader_and_node_list)
+        # There are some local readers which are the same worker process as the writer.
```

Comment:
move to L96?

Reply:
update: https://github.com/ray-project/ray/pull/49213/commits/b4c7e2641102167a7df166c04e7d82c203099aa2